src/holon/services/costs_table.py

```diff
@@ -0,0 +1,180 @@
+
+    @classmethod
+    def from_al_output(cls, al_output, scenario):
+        actors = ActorWrapper.from_scenario(scenario)
+        return cls([CostItem.from_dict(item, actors) for item in al_output])
+
+
+class CostTable:
+    def __init__(self, cost_items, use_subgroup=None) -> None:
+        """cost_items is a list of CostItems"""
+        self._use_subgroup = use_subgroup
+        self.table = cost_items
+
+    @property
+    def table(self):
+        return self._table
+
+    @table.setter
+    def table(self, cost_items):
+        self._table = {}
+        for item in cost_items:
+            self.__add_to_table(item)
+        self.__fill_out_table()
+
+    def __add_to_table(self, item):
+        """Adds the item to the table"""
+        try:
+            self._table[self.__name_from(item)][self.__name_to(item)] += item.price
+        except:
+            self.__add_to_group(item)
+
+    def __add_to_group(self, item):
+        try:
+            self._table[self.__name_from(item)][self.__name_to(item)] = item.price
+        except KeyError:
+            self.__add_from_group(item)
+
+    def __add_from_group(self, item):
+        """
+        Also needs to add self as None
+        TODO: move some functionality from fill_out_table here
+        """
+        self._table[self.__name_from(item)] = {
+            self.__name_to(item): item.price,
+            self.__name_from(item): None,
+        }
+
+    def __fill_out_table(self):
+        # we can also keep a global set in memory (self) where we add to in __add_from_group
+        all_groups = set((key for value in self.table.values() for key in value.keys()))
+        basic = {key: 0.0 for key in all_groups}
+        for group in all_groups:
+            self._table[group] = basic | self._table.get(group, {})
+
+    def __name_from(self, item):
+        return (
+            item.from_subgroup() if self._use_subgroup == item.from_group() else item.from_group()
+        )
+
+    def __name_to(self, item):
+        return item.to_subgroup() if self._use_subgroup == item.to_group() else item.to_group()
+
+
+class ActorWrapper:
+    def __init__(self, actors) -> None:
+        """Where actors is the Django equivalent of AR relation of Actors of the scenario"""
+        self.actors = actors
+
+    def find(self, actor_name):
+        """
+        Strips the AL prefix from the actor name and returns the corresponding Actor
+        """
+        return self.actors.get(id=int(actor_name[3:]))
+
+    @classmethod
+    def from_scenario(cls, scenario):
+        return cls(scenario.actor_set)
+
+
+class CostItem:
+    """Represents one contract in the AL output"""
+
+    def __init__(self, to_actor, from_actor, price) -> None:
+        self.from_actor = from_actor
+        self.to_actor = to_actor
+        self.price = price
+        print(f"CostItem: {self.from_group()} -> {self.to_group()} = {self.price}")
```

Comment:
Can you remove the (debugging) print statements as it can slow down the process and is considered bad practice in production code? Important stuff can instead be sent to a logger 🙂

Reply:
Whoopsie, must have been through the rebase. I'm sure I had removed them 🥷, thanks for the headsup!